pkg/operator/operator.go

```diff
@@ -743,3 +776,244 @@ func (o *Operator) Config(ctx context.Context, key string) (*manifests.Config, e
 
 	return c, nil
 }
+
+// Upgradeable verifies whether the operator can be upgraded or not. It returns
+// the ConditionStatus with optional reason and message.
+func (o *Operator) Upgradeable(ctx context.Context) (configv1.ConditionStatus, string, string, error) {
+	if !o.lastKnowInfrastructureConfig.HighlyAvailableInfrastructure() {
+		return configv1.ConditionTrue, "", "", nil
+	}
+
+	workloadsCorrectlySpread, reason, message, err := o.WorkloadsCorrectlySpread(ctx)
+	if err != nil {
+		return configv1.ConditionUnknown, "", "", err
+	}
+
+	if !workloadsCorrectlySpread {
+		return configv1.ConditionFalse, reason, message, nil
+	}
+
+	return configv1.ConditionTrue, reason, message, nil
+}
+
+func (o *Operator) ensureNodesAreUncordonned(ctx context.Context) error {
```

Comment:
When cordoning a node, CMO could set an annotation on the node resource (e.g. `openshift.io/cluster-monitoring-cordoned="node marked as unschedulable by cluster-monitoring-operator to reschedule ns/pod on another node"`). Then  `ensureNodesAreUncordonned()` would get all nodes with `spec.unschedulable=true` (this is a supported field selector) and uncordon node(s) that have the CMO annotation. Then we wouldn't need to maintain a map of nodes to uncordon and we would be resilient to crashes/restarts of CMO.

Reply:
Makes complete sense, I updated the logic.